### PR TITLE
Remove CPU limits from gardener components

### PR DESCRIPTION
**Pull request by stoyanr**
**How to categorize this PR?**

/kind enhancement

**What this PR does / why we need it**:
* Removes CPU limits from all components to prevent CPU throttling due to reaching limits.

The following will be added in separate PRs:
* Remove or adjust memory limits from all components to match actual usage on landscapes.
* Fix `VerticalPodAutoscaler` resources to use [RequestsOnly instead of RequestsAndLimits](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L209-L215)
* Fix `VerticalPodAutoscaler` and `Hvpa` resources to remove `MaxAllowed` where it's present

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```noteworthy operator
CPU limits from all gardener components have been removed to prevent CPU throttling due to reaching limits.
```

### CODE REVIEW

**Comment by stoyanr** [comment]
I consider this PR ready for review now.

**Review by rfranzke — DISMISSED** [review]
/lgtm

File: charts/gardener/controlplane/values.yaml
Comment (ialidzhikov) [inline charts/gardener/controlplane/values.yaml:349]: For GCM there was a `automaxprocs` handling introduced by the following PR https://github.com/gardener/gardener/pull/4979. If we understand correctly it is automatically set `GOMAXPROCS` to match Linux container CPU quota. If we now completely drop the CPU limit for the GCM, do we still benefit/need this `go.uber.org/automaxprocs` library usage in GCM?

cc @timebertt @wyb1 @istvanballok

File: charts/istio/istio-ingress/templates/deployment.yaml
Comment (ialidzhikov) [inline charts/istio/istio-ingress/templates/deployment.yaml:83]: @dguendisch asked us to do not rollout the istio-ingressgateway Pods until the following issue is resolved https://github.com/gardener/machine-controller-manager/issues/679 - ref https://github.com/gardener/gardener/pull/5468#issuecomment-1048937877

@dguendisch is it still the case?

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (ialidzhikov) [inline pkg/operation/botanist/component/etcd/etcd.go]: I think we pay attention to etcd StatefulSet changes and we try to do not rollout etcd-main when no critical change is present - simply because we don't have yet multi-node etcd-main and each etcd-main rollout leads to downtime.

I guess etcd-druid soon have a new release (that is supposed to rollout etcd-main Pods?) and I would vote to sync and combine both changes in single gardener release.

cc @rfranzke @timebertt (as the v1.44 release responsible)

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (ialidzhikov) [inline pkg/operation/botanist/component/etcd/etcd.go]:
```suggestion
				// TODO(stoyanr): This limit can't be removed currently because etcd-druid supplies a default of 100m.
				// Use a very high value instead and remove it later after <issue-link> is resolved.
```

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (ialidzhikov) [inline pkg/operation/botanist/component/etcd/etcd.go]:
```suggestion
				// TODO(stoyanr): This limit can't be removed currently because etcd-druid supplies a default of 100m.
				// Use a very high value instead and remove it later after <issue-link> is resolved.
```

File: charts/gardener/provider-local/internal/machine-controller-manager/seed/templates/deployment.yaml
Comment (ialidzhikov) [inline charts/gardener/provider-local/internal/machine-controller-manager/seed/templates/deployment.yaml:124]: We see that a cpu limit is not removed from `./pkg/provider-local/controller/network/actuator.go`. Is it because the corresponding container is with guaranteed QoS? And, a question in general, do we also remove the CPU limits from guaranteed QoS Pods? If not, I guess it would makes sense to document this handling in a developer guide.

File: charts/istio/istio-ingress/templates/deployment.yaml
Comment (ialidzhikov) [inline charts/istio/istio-ingress/templates/deployment.yaml:83]: A comment in general - are we confident enough that this change is safe and won't lead to unexpected behaviour? If we are not, does it makes sense to put it behind a feature gate to give Gardener operators control to rollout and rollback (if needed).

**Review by ialidzhikov — DISMISSED** [review]
We had a look with @plkokanov and we listed few concerns inline.

File: charts/istio/istio-ingress/templates/deployment.yaml
Comment (dguendisch) [reply charts/istio/istio-ingress/templates/deployment.yaml:83]: I think we won't find anything more on the istio and following melt-down. So given that https://github.com/gardener/gardener/pull/5497 is already deployed, it should protect enough from any melt-down when istio is having troubles. So I'd think that it's good to go now. Do you agree @ashwani2k and @ScheererJ ?

Comment (ialidzhikov) [reply charts/istio/istio-ingress/templates/deployment.yaml:83]: If we are good to go then we should unhold https://github.com/gardener/gardener/pull/5340 and https://github.com/gardener/gardener/pull/5468 as well. @ashwani2k , @himanshu-kun @ScheererJ please confirm.

File: charts/gardener/controlplane/values.yaml
Comment (stoyanr) [reply charts/gardener/controlplane/values.yaml:349]: How exactly is this related to the limits? "Linux container CPU quota" is not determined by the limits, or perhaps I don't understand what you mean?

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (stoyanr) [reply pkg/operation/botanist/component/etcd/etcd.go]: Makes sense, thanks!

File: charts/istio/istio-ingress/templates/deployment.yaml
Comment (stoyanr) [reply charts/istio/istio-ingress/templates/deployment.yaml:83]: I am not sure how I could put a change like this behind a feature gate - it contains a few dozen small changes, many of them to Helm template files. 
Even if it's possible somehow, I would strongly vote not to complicate our lives so much. I am personally rather confident that this change will not have any really serious negative consequences in general (nothing will be "broken"), and if it does, we would be able to see it dev / staging.

File: charts/gardener/provider-local/internal/machine-controller-manager/seed/templates/deployment.yaml
Comment (stoyanr) [reply charts/gardener/provider-local/internal/machine-controller-manager/seed/templates/deployment.yaml:124]: The idea was to not remove CPU limits from guaranteed QoS pods (since they would not be guaranteed anymore). Did you notice such changes? What exactly do you want documented in the developer guide? From my perspective, it is self-evident that a pod with a guaranteed QoS should have limits (since guaranteed means requests == limits).

File: charts/gardener/controlplane/values.yaml
Comment (istvanballok) [reply charts/gardener/controlplane/values.yaml:349]: > If we now completely drop the CPU limit for the GCM, do we still benefit/need this go.uber.org/automaxprocs library usage in GCM?

Yes, you are right - if there are no limits, i.e. there is no cgroup CPU quota (which is derived from the k8s limits), the automaxprocs feature will not give any benefits (and not cause any harm), the go runtime will use all the CPU cores of the host to execute go routines, inline with the aims of this PR. [1](https://github.com/uber-go/automaxprocs/blob/beef36874df7a16ce9ba64a31e0607f8b2cecd7b/internal/runtime/cpu_quota_linux.go#L42) [2](https://github.com/uber-go/automaxprocs/blob/beef36874df7a16ce9ba64a31e0607f8b2cecd7b/maxprocs/maxprocs.go#L111)

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (stoyanr) [reply pkg/operation/botanist/component/etcd/etcd.go]: Done

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (stoyanr) [reply pkg/operation/botanist/component/etcd/etcd.go]: Done

File: charts/istio/istio-ingress/templates/deployment.yaml
Comment (ScheererJ) [reply charts/istio/istio-ingress/templates/deployment.yaml:83]: @ialidzhikov I agree with @dguendisch.

#5468 should be ready. #5340 needs more work from my side.

Comment (ialidzhikov) [reply charts/istio/istio-ingress/templates/deployment.yaml:83]: Sure, I agree that it can be overhead. Just wanted to put this option on the table (in case we have doubts) as I was not able to completely follow the communication/motivation and all internals behind the CPU limits removal. Then let's do it without a feature gate.

File: charts/gardener/provider-local/internal/machine-controller-manager/seed/templates/deployment.yaml
Comment (ialidzhikov) [reply charts/gardener/provider-local/internal/machine-controller-manager/seed/templates/deployment.yaml:124]: > Did you notice such changes?

I didn't but looks like you did - https://github.com/gardener/gardener/compare/8226c5dff21f937c53ab9a7972e6b2e221e111aa..f9062639169187226bc4d4d3a75213c861c34bff. Thanks!

> From my perspective, it is self-evident that a pod with a guaranteed QoS should have limits (since guaranteed means requests == limits).

Hm, it was not self-evident for me and @plkokanov . But it can be the case just because we miss context on the discussions. IMO it would make sense to put somewhere few sentences for developers/maintainers to do not add a CPU limits for a Burstable Pod (with the appropriate reasoning). Otherwise we risk the rule imposed by this PR to be violated for a new component that gets introduced tomorrow.

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (ialidzhikov) [reply pkg/operation/botanist/component/etcd/etcd.go]: Thanks for the related PR!
Nit: Can we also add an owner to the TODO to make sure that we follow-up on it relatively soon?

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (ialidzhikov) [reply pkg/operation/botanist/component/etcd/etcd.go]: Ok, then let's add the
/hold
label until we have the new etcd-druid version in.

Just to make sure, @rfranzke you mentioned that the new etcd-druid will anyways rollout the etcd-main Pods, right?

File: charts/gardener/controlplane/values.yaml
Comment (ialidzhikov) [reply charts/gardener/controlplane/values.yaml:349]: Thanks, @istvanballok ! Then let's consider dropping the automaxprocs feature as a separate follow up? It will be still possible to specify cpu limits when applying the gardener control plane chart and automaxprocs feature can be useful then but I rather understand that we want to run without cpu limits. Anyways, we only enabled this for GCM and didn't manage to follow-up with other components. WDYT?

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (rfranzke) [reply pkg/operation/botanist/component/etcd/etcd.go]: Yes, it will. @stoyanr maybe you can carve out the etcd-related changes into a separate PR so that we don't have to hold the entire rest until the new druid version arrives.

File: charts/gardener/provider-local/internal/machine-controller-manager/seed/templates/deployment.yaml
Comment (stoyanr) [reply charts/gardener/provider-local/internal/machine-controller-manager/seed/templates/deployment.yaml:124]: This PR is not intended to impose any rule. In the discussion that we had the consensus is as follows. There is nothing wrong with having limits in general. If a component owner knows a good limit (e.g. because he knows the code), it's perfectly ok to have one. Our current issue with limits is that they are mostly based on nothing, and therefore hurt much more than help. 

Therefore, I will not document that there should be no CPU limits, because I myself don't think this is the case, and neither do the participants in the discussion.

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (stoyanr) [reply pkg/operation/botanist/component/etcd/etcd.go]: I will do it anyway without the nit.

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (stoyanr) [reply pkg/operation/botanist/component/etcd/etcd.go]: When is the new etcd-druid version supposed to arrive? How hard it is to release and include it in the current g/g release, so that I will not have to spend even more effort on this PR? Can I perhaps help with that, instead of splitting?

Comment (rfranzke) [reply pkg/operation/botanist/component/etcd/etcd.go]: Well, it should have been released this week, but there are still relevant open PRs which need to be merged first (before the team conducts testing, I think). You can sync with @ashwani2k and @ishan16696 about it.

File: charts/gardener/controlplane/values.yaml
Comment (istvanballok) [reply charts/gardener/controlplane/values.yaml:349]: Hi @ialidzhikov (@wyb1 cc)
I do not have a strong opinion
- we could revert https://github.com/gardener/gardener/pull/4979 as a cleanup action
- the automaxprocs feature shouldn't do harm even without limits, and if we'll happen to have limits later, I think it is meaningful to let the library reduce the number of go routine executing threads. Note that there are some caveats here as well (floor, ceil) and this topic of CPU limits and golang is not fully understood (by me) yet. Now that we are moving away from limits, we could either wait a bit and (later or now) do the cleanup action.

File: charts/istio/istio-ingress/templates/deployment.yaml
Comment (himanshu-kun) [reply charts/istio/istio-ingress/templates/deployment.yaml:83]: yes from my side.

**Comment by vlerenc** [comment]
Thank you @stoyanr and all for reviewing it. I also do not see any component in the list where we actually would want throttling for (I would think, this makes only sense for batch/job processing and we don't have such components much/none here if I am not mistaken).

/lgtm

A detail question though about https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/kubeapiserver.go that I saw in this PR: What is it about? What is that `scalingClass`? Is that really dependent on the nodes/statically? I must have forgotten, but what's the background?

Follow-up question: Is that possibly, in another PR, a way to curtail non-production clusters (actually limit them/everything that is not labeled with the production purpose)?

**Comment by vlerenc** [comment]
cc @rfranzke as well...

A detail question though about https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/kubeapiserver.go that I saw in this PR: What is it about? What is that scalingClass? Is that really dependent on the nodes/statically? I must have forgotten, but what's the background?

Follow-up question: Is that possibly, in another PR, a way to curtail non-production clusters (actually limit them/everything that is not labeled with the production purpose)?

**Comment by rfranzke** [comment]
> A detail question though about https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/kubeapiserver.go that I saw in this PR: What is it about? What is that scalingClass? Is that really dependent on the nodes/statically? I must have forgotten, but what's the background?

@vlerenc This was implement as part of https://github.com/gardener/gardener/issues/255 with this PR: https://github.com/gardener/gardener/pull/315. I don't know the details, but I see that you yourself were heavily involved in the discussions 😉

**Comment by vlerenc** [comment]
LOL, thanks @rfranzke.

(1) https://github.com/gardener/gardener/issues/255 was a band-aid from mid 2018 for the poor VPA v1 for the actual ticket...

(2) https://github.com/gardener/gardener/issues/16 that we couldn't implement with the former VPA v1 and so we did this, because we hit...

(3) https://github.com/gardener/gardener/issues/79, where KCM stopped working under load

cc @voelzmo we might want to rid is of all that `scalingClass` workarounds now with VPA v2, if we can make it work at scale and as we want it to work.

So funny/shocking how long this topic is now with us. I see Gaurav made the changes. He was the first of the 3 colleagues on the auto-scaling topic - not in parallel - in sequence over 4 years. :-)

Time to give it a new shot and improve or ideally get rid of all these hacks.

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (rfranzke) [reply pkg/operation/botanist/component/etcd/etcd.go]: There is a PR for the new etcd-druid release: https://github.com/gardener/gardener/pull/5693

**Comment by stoyanr** [comment]
> Time to give it a new shot and improve or ideally get rid of all these hacks.

@vlerenc I will take a look, and perhaps improve it in a separate PR.

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (stoyanr) [reply pkg/operation/botanist/component/etcd/etcd.go]: @rfranzke #5693 is now open (and mergeable), I assume this means we could close this discussion and unhold the PR?

Comment (rfranzke) [reply pkg/operation/botanist/component/etcd/etcd.go]: Yes, let's wait for this upgrade PR to be merged (soon/today) and then we can also merge your PR here.

Comment (stoyanr) [reply pkg/operation/botanist/component/etcd/etcd.go]: In fact, since https://github.com/gardener/etcd-druid/pull/312 has also been included in the new `etcd-druid` release, I could get rid of the TODOs for `Etcd` resource limits once it's merged. Do you think it will be merged soon?

Comment (rfranzke) [reply pkg/operation/botanist/component/etcd/etcd.go]: Yes, I think so. Please get rid of the TODOs :)

**Comment by vlerenc** [comment]
Thanks @stoyanr. I am just a bit cautious here as well.

The point is that it may still be necessary to have "decent" `requests`, "oriented" at the number/size of the nodes, so that we start in the correct ballpark and avoid lengthy up-scaling iterations with also the current VPAv2 (unless tuned). I say that because I remember many years before that e.g. waking up clusters (after hibernation) took very long to scale up and become operational again. Scaling classes "catapulted" control planes into the right ballpark. As long as VPAv2 is making poor recommendations or isn't bumping up faster, most of that code would probably still be required. We can only really rid ourselves of it once VPAv2 overcomes it current problems. At least, that's my gut feeling.

**Comment by stoyanr** [comment]
> As long as VPAv2 is making poor recommendations or isn't bumping up faster, most of that code would probably still be required. We can only really rid ourselves of it once VPAv2 overcomes it current problems. 

OK, in this case I will postpone this for later.

**Comment by stoyanr** [comment]
@vlerenc I examined the `scalingClass` implementation and played with it a bit, it seems that when HVPA is enabled (it's a feature gate in gardenlet that I believe is enabled by default), the scaling class is only really taken into account the first time the deployment is created. If the deployment already exists, the resources are copied from the existing deployment. This means that if the cluster started with a low initial number of nodes, increasing the number of nodes later will have no effect on the resources in the deployment. Hibernating / waking up will certainly have no effect either.

I am rather skeptical of this complex implementation (scaling class + HVPA), but as agreed, let's take another look and possibly change it once we are more confident of VPA recommendations.

**Comment by vlerenc** [comment]
@stoyanr My memory is from many years back with end users complaining in Slack and whatever you have found out today counts (not my memory), because we may have influenced the behaviour directly or indirectly many times over since 2018.

File: pkg/operation/botanist/component/etcd/etcd.go
Comment (stoyanr) [reply pkg/operation/botanist/component/etcd/etcd.go]: The above PR is merged now, closing this discussion.

Comment (rfranzke) [reply pkg/operation/botanist/component/etcd/etcd.go]: /hold cancel

**Comment by stoyanr** [comment]
@vlerenc @rfranzke The new `etcd-druid` release has been merged and I removed the few TODOs in the PR (now possible with the new `etcd-druid` version). I consider the PR ready to be merged. Could I get an approval or do you have any other open points?

**Review by rfranzke — COMMENTED** [review]
/lgtm
/approve
/hold

@vlerenc any objections? If not please `/hold cancel` so that the PR gets auto-merged by Prow.

**Comment by vlerenc** [comment]
/hold cancel
@stoyanr Thanks for the PR. 👍 Let's roll it our carefully now, watching for possible issues, increased node usage, etc. please.
@rfranzke I have not `hold` this one (in fact, I `lgtm`'d it, because that's what we want - I can only check from the "intention" level and not every line anyway/am no expert), but you wanted to double-check because of Prow's automation?

**Comment by rfranzke** [comment]
I simply wasn't sure whether you are fine merging this PR now because there were quite some follow-up questions after your `/lgtm` and I couldn't follow everything in detail :)

**Pull request merged**
